Update qengine flag in python to string

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26620 Update qengine flag in python to string**

Summary: This change updates torch.backend.quantized.engine to accept string ("fbgemm"/"qnnpack"/"none" for now).
set_qengine and get_qengine return an int which represents the at::QEngine enum

Test Plan:
python test/test_torch.py

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17533582](https://our.internmc.facebook.com/intern/diff/D17533582)

supriyar committed:
Update qengine flag in python to string

Summary: This change updates torch.backend.quantized.engine to accept string ("fbgemm"/"qnnpack"/"none" for now).
set_qengine and get_qengine return an int which represents the at::QEngine enum

Test Plan:
python test/test_torch.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Update qengine flag in python to string"


Summary: This change updates torch.backend.quantized.engine to accept string ("fbgemm"/"qnnpack"/"none" for now).
set_qengine and get_qengine return an int which represents the at::QEngine enum

Test Plan:
python test/test_torch.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/test_torch.py:
```diff
@@ -2115,12 +2115,11 @@ def test_inference(default_dtype):
-        #     torch._C._set_qengine(qe)
-        #     assert torch._C._get_qengine() == qe, 'qengine not set successfully'
-        # torch._C._set_qengine(original_qe)
+        original_qe = torch._C._get_qengine()
```
actually, looks like user is going to use `toch.backends.quantized.engine = 'qnnpack'` to set the engine? if so we should do this here instead.

jerryzh168 reviewed (changes requested):
had a question about user API

supriyar committed:
Update on "Update qengine flag in python to string"


Summary: This change updates torch.backend.quantized.engine to accept string ("fbgemm"/"qnnpack"/"none" for now).
set_qengine and get_qengine return an int which represents the at::QEngine enum

Test Plan:
python test/test_torch.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/test_torch.py:
```diff
@@ -2115,12 +2115,12 @@ def test_inference(default_dtype):
-        #     assert torch._C._get_qengine() == qe, 'qengine not set successfully'
-        # torch._C._set_qengine(original_qe)
+        original_qe = torch.backends.quantized.engine
+        all_engines = {0 : 'none', 1 : 'fbgemm', 2 : 'qnnpack'}
```
should we make this as a global helper function so that users can use it too? I think we can have a `get_qengine_str` in `torch/backends/quantized/__init__.py ` and call this function here.

supriyar replied:
Sounds good.

jerryzh168 commented on test/test_torch.py:
```diff
@@ -2115,12 +2115,12 @@ def test_inference(default_dtype):
+        for qe in qengines:
+            torch.backends.quantized.engine = all_engines.get(qe)
+            assert torch.backends.quantized.engine == qe, 'qengine not set successfully'
+        torch._C._set_qengine(original_qe)
```
this should be changed to `torch.backends.quantized.engine = original_qe`

jerryzh168 commented on test/common_quantized.py:
```diff
@@ -65,8 +65,12 @@ def _calculate_dynamic_qparams(X, dtype):
         yield
     finally:
-        torch.backends.quantized.engine = torch.fbgemm
+        qengines = torch._C._supported_qengines()
```
should we have another wrapper around `_supported_qengines` which returns a list of strings to user?

supriyar replied:
Yeah, it will make everything string which is more consistent.

supriyar committed:
Update on "Update qengine flag in python to string"


Summary: This change updates torch.backend.quantized.engine to accept string ("fbgemm"/"qnnpack"/"none" for now).
set_qengine and get_qengine return an int which represents the at::QEngine enum

Test Plan:
python test/test_torch.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (approved):
LGTM Thanks!

jerryzh168 commented on test/common_quantized.py:
```diff
@@ -65,8 +65,12 @@ def _calculate_dynamic_qparams(X, dtype):
     finally:
-        torch.backends.quantized.engine = torch.fbgemm
+        qengines = torch._C._supported_qengines()
+        if 2 in qengines:
```
oh, you will need to change this as well

supriyar replied:
oops, missed this.

supriyar committed:
Update on "Update qengine flag in python to string"


Summary: This change updates torch.backend.quantized.engine to accept string ("fbgemm"/"qnnpack"/"none" for now).
set_qengine and get_qengine return an int which represents the at::QEngine enum

Test Plan:
python test/test_torch.py

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
@pytorchbot retest this please

dzhulgakov commented:
We could have passes string all the way to C++ to avoid enum duplication (and it's much easier to handle strings if you use pybind11). But it's fine to do later.
